### PR TITLE
Fiat currency; fixes for running on Apple Silicon

### PR DESCRIPTION
**Pull request by tw0po1nt**
This code review checklist is intended to serve as a starting point for the author and reviewer, although it may not be appropriate for all types of changes (e.g. fixing a spelling typo in documentation).  For more in-depth discussion of how we think about code review, please see [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md).

# Author

- [x] Self-review: Did you review your own code in GitHub's web interface? Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs.
- [x] Does the code abide by the [Coding Guidelines](../blob/main/docs/CODING_GUIDELINES.md)?
- [ ] Automated tests: Did you add appropriate automated tests for any code changes?
- [ ] Code coverage: Did you check the code coverage report for the automated tests?  While we are not looking for perfect coverage, the tool can point out potential cases that have been missed.
- [ ] Documentation: Did you update Docs as appropiate? (E.g [README.md](../blob/main/README.md), etc.)
- [x] Run the app: Did you run the app and try the changes? 
- [ ] Did you provide Screenshots of what the App looks like before and after your changes as part of the description of this PR? (only applicable to UI Changes)
- [x] Rebase and squash: Did you pull in the latest changes from the main branch and squash your commits before assigning a reviewer? Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit.

# Reviewer

- [ ] Checklist review: Did you go through the code with the [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md) checklist?
- [ ] Ad hoc review: Did you perform an ad hoc review?  _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
- [ ] Automated tests: Did you review the automated tests?
- [ ] Manual tests: Did you review the manual tests?_You will find manual testing guidelines under our [manual testing section](../blob/main/docs/testing/manual_testing)_
- [ ] How is Code Coverage affected by this PR? _We encourage you to compare coverage before and after your changes and when possible, leave it in a better place. [Learn More...](../blob/main/docs/testing/local_coverage.md)_
- [ ] Documentation: Did you review Docs, [README.md](../blob/main/README.md), [LICENSE.md](../blob/main/LICENSE.md), and [Architecture.md](../blob/main/docs/Architecture.md) as appropriate?
- [ ] Run the app: Did you run the app and try the changes? While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data.

### CODE REVIEW

**Review by tw0po1nt — COMMENTED** [review]
Self-review completed

File: modules/Sources/Utils/NumberFormatter+zcash.swift [hunk 1]
```diff
@@ -8,6 +8,14 @@
 import Foundation
 
 extension NumberFormatter {
+    public static let currencyFormatter: NumberFormatter = {
+        var formatter = NumberFormatter()
+        formatter.maximumFractionDigits = 2
+        formatter.numberStyle = .decimal
```
Comment (pacu) [inline modules/Sources/Utils/NumberFormatter+zcash.swift:15]: [Non-Blocking] I wonder if it's better to use `.currency` instead.

File: modules/Sources/Features/Home/Home.swift [hunk 2]
```diff
@@ -163,10 +180,22 @@ public struct Home: Reducer {
...
                 state.destination = .alert(.cantStartSync(error))
                 return .none
+            case .fetchLatestFiatPrice:
+                guard !state.isFiatLoaded, state.preferredCurrency != . off else { return .none }
+                state.isFiatLoaded = true
+                return .run { [preferredCurrency = state.preferredCurrency] send in
+                    try? await send(.latestFiatResponse(fiatPriceClient.getZcashPrice(preferredCurrency)))
+                }
+            case let .latestFiatResponse(price):
+                userStoredPreferences.setLatestFiatPrice(price)
```
Comment (pacu) [inline modules/Sources/Features/Home/Home.swift]: [Blocking] 
is this a "caching" workaround?

How would you determine that the value is not fresh anymore?

File: modules/Sources/Features/Home/NighthawkSettings/Fiat/FiatView.swift [hunk 3]
```diff
@@ -6,19 +6,90 @@
...
+                        .multilineTextAlignment(.leading)
+                        .lineSpacing(6)
+                    
+                    RadioSelectionList(
+                        options: NighthawkSetting.FiatCurrency.allCases,
+                        selection: viewStore.$selectedFiatCurrency.animation(.none)
+                    ) { option in
+                        HStack {
+                            switch option {
+                            case .usd:
```
Comment (pacu) [inline modules/Sources/Features/Home/NighthawkSettings/Fiat/FiatView.swift]: [non-blocking] I would move this out of the view to a helper function, to resolve the Localization key and leave the view more readable

File: modules/Sources/Dependencies/FiatPriceClient/FiatPriceClient.swift [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+extension FiatPriceClient: DependencyKey {
+    struct FiatPriceApiResponse: Codable, Equatable {
+        let data: [String: Double]
+        
+        enum CodingKeys: String, CodingKey {
+            case data = "zcash"
+        }
+    }
+    
+    public static let liveValue = Self(
```
Comment (pacu) [inline modules/Sources/Dependencies/FiatPriceClient/FiatPriceClient.swift:26]: I'd consider wrapping this up into a caching client that can determine when it is appropriate to refresh the value or return what's already cached. You could also add a setting in your awesome settings screen to adjust the "refresh" time, so that you can let the user decide how often to refresh the cache.

File: modules/Sources/Dependencies/FiatPriceClient/FiatPriceClient.swift [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+            let simplePriceUrl = URL.coinGeckoApi
+                .appending(path: "simple/price")
+                .appending(
+                    queryItems: [
+                        .init(name: "ids", value: "zcash"),
+                        .init(name: "vs_currencies", value: currency.rawValue)
+                    ]
+                )
+            
+            let (data, _) = try await URLSession.shared.data(from: simplePriceUrl)
```
Comment (pacu) [inline modules/Sources/Dependencies/FiatPriceClient/FiatPriceClient.swift:39]: [Blocking] I have a concern about the timing of these calls. If the app polls the value in a predictable way, this fiat conversion is an excellent vector for timing attacks from an adversary. The pace of a service like this can be used to  profile a user an identify it as a NH user and use this also as a way of knowing when a target has started using the app.

This has a sort of simple solution that is randomizing how this value is refreshed over a range of time, so that it can't be statistically determined.

**Review by pacu — CHANGES_REQUESTED** [review]
Great PR! awesome feature!.

I have left blocking and non-blocking comments that might be resolved with code, clarifying comments or TODOs with future issues to be resolved later.

File: modules/Sources/Dependencies/FiatPriceClient/FiatPriceClient.swift (hunk 5 above)
Comment (nighthawk24) [reply modules/Sources/Dependencies/FiatPriceClient/FiatPriceClient.swift:39]: By default, fiat currency API call is disabled in Nighthawk, aligning with our security-first approach to third-party services. Users may decide to opt-in to this feature, ensuring protection against possible, predictable timing-based attacks.

When a Nighthawk user chooses to enable fiat currency display, the price API is only polled at the start of the app session, typically lasting 5-15 minutes. The term 'using' the app refers to users broadcasting a transaction, not just opening the app. Users may open the app for various purposes, such as sharing their receiving address or viewing transaction history, which doesn't convert to user-tracking due to the open nature of the pricing API used by wallets in the cryptocurrency ecosystem.

This approach distinguishes Nighthawk from other wallets that constantly poll real-time data during public interactions with the blockchain, reducing the risk of user tracking. However, randomizing refreshed values will result in showing inaccurate or outdated fiat currency rates to the user. This is a tradeoff for users that decide to opt-in to viewing ZEC price within the app.

Users are also notified of the following text before they decide to opt-in to this feature:

"Choose your local currency so that we can show you a close estimate of how much your ZEC is worth.

When you first open the app, the exchange rate is fetched from CoinGecko API v3 and used to calculate values. Your fund amounts are never revealed to any server."

NOTE: Please report any possible vulnerabilities or attack vectors following our disclosure policy in the project's README.

**Comment by nighthawk24** [comment]
> Great PR! awesome feature!.
> 
> 
> 
> I have left blocking and non-blocking comments that might be resolved with code, clarifying comments or TODOs with future issues to be resolved later. 

Thank you, @pacu, for your invaluable and timely code reviews, especially as we approach Nighthawk's public release.

However, it's important to note that the 2023 grants committee has unanimously voted to reject the wallet maintenance proposal, following an already partially approved budget for our team. This decision poses challenges for Nighthawk developers ability to meeting user needs. We encourage you to contribute by opening a pull request to the project if you have ideas or solutions.

https://forum.zcashcommunity.com/t/nighthawk-wallet-maintenance-and-self-funding-improvements/45048/16

File: modules/Sources/Utils/NumberFormatter+zcash.swift (hunk 1 above)
Comment (tw0po1nt) [reply modules/Sources/Utils/NumberFormatter+zcash.swift:15]: .currency adds the symbol, such as '$' or '€', but with our designs we want something like "around 2.50 USD", so that's why I chose this approach.

File: modules/Sources/Features/Home/Home.swift (hunk 2 above)
Comment (tw0po1nt) [reply modules/Sources/Features/Home/Home.swift]: It wasn't really so much as a "caching" scheme as it was a "shared global" scheme so I don't have to pass it around everywhere. But in chat with Adi, we really want this to be tied to the app session, so this should truly just be stored in non-persistent memory. Will change to resolve this.

File: modules/Sources/Dependencies/FiatPriceClient/FiatPriceClient.swift (hunk 4 above)
Comment (tw0po1nt) [reply modules/Sources/Dependencies/FiatPriceClient/FiatPriceClient.swift:26]: See my comment above. Normally you'd be right and yes, we'd have some sort of intelligent caching scheme, but in this case, we just want to always fetch this value if it isn't present in-mem, and discard it when the app's memory is reclaimed.

File: modules/Sources/Dependencies/FiatPriceClient/FiatPriceClient.swift (hunk 5 above)
Comment (pacu) [reply modules/Sources/Dependencies/FiatPriceClient/FiatPriceClient.swift:39]: I consider this comment resolves my comment. Thanks @nighthawk24

File: modules/Sources/Dependencies/SDKSynchronizer/SDKSynchronizerLive.swift [hunk 6]
```diff
@@ -29,7 +29,7 @@ extension SDKSynchronizerClient {
             spendParamsURL: databaseFiles.spendParamsURLFor(network),
             outputParamsURL: databaseFiles.outputParamsURLFor(network),
             saplingParamsSourceURL: SaplingParamsSourceURL.default,
-            loggingPolicy: .noLogging
+            loggingPolicy: .default(.debug)
```
Comment (pacu) [inline modules/Sources/Dependencies/SDKSynchronizer/SDKSynchronizerLive.swift]: [non-blocking] Is this change intentional?

Comment (tw0po1nt) [reply modules/Sources/Dependencies/SDKSynchronizer/SDKSynchronizerLive.swift]: It is not, thanks for the catch.

**Comment by pacu** [comment]
> > Great PR! awesome feature!.
> > I have left blocking and non-blocking comments that might be resolved with code, clarifying comments or TODOs with future issues to be resolved later.
> 
> Thank you, @pacu, for your invaluable and timely code reviews, especially as we approach Nighthawk's public release.
> 
> However, it's important to note that the 2023 grants committee has unanimously voted to reject the wallet maintenance proposal, following an already partially approved budget for our team. This decision poses challenges for Nighthawk developers ability to meeting user needs. We encourage you to contribute by opening a pull request to the project if you have ideas or solutions.
> 
> https://forum.zcashcommunity.com/t/nighthawk-wallet-maintenance-and-self-funding-improvements/45048/16

I'm sorry that you folks didn't get that grant :(. I will continue contributing to NH's codebase for sure!

**Review by pacu — DISMISSED** [review]
thank you for taking in my feedback!

utACK

**Review by nighthawk24 — DISMISSED** [review]

**Review by pacu — APPROVED** [review]
utACK
